lib/icmp/icmpv4_parse.ml

```diff
@@ -17,25 +18,34 @@ type t = {
 let subheader_of_cstruct ty buf =
   let open Cstruct.BE in
   match ty with
-  | 0 | 8 | 13 | 14 | 15 | 16 -> Id_and_seq (get_uint16 buf 0, get_uint16 buf 2)
-  | 3 | 11 | 4 -> Unused
-  | 5 -> Address (Ipaddr.V4.of_int32 (get_uint32 buf 0))
-  | 12 -> Pointer (Cstruct.get_uint8 buf 0)
-  | _ -> Unused
+  | Echo_request | Echo_reply
+  | Timestamp_request | Timestamp_reply
+  | Information_request | Information_reply ->
+    Id_and_seq (get_uint16 buf 0, get_uint16 buf 2)
+  | Destination_unreachable -> Next_hop_mtu (get_uint16 buf 2)
+  | Time_exceeded
+  | Source_quench -> Unused
+  | Redirect -> Address (Ipaddr.V4.of_int32 (get_uint32 buf 0))
+  | Parameter_problem -> Pointer (Cstruct.get_uint8 buf 0)
 
 let input buf =
-  if Cstruct.len buf < Icmpv4_wire.sizeof_icmpv4 then
+  let open Rresult in
+  let check_len () =
+  if Cstruct.len buf < sizeof_icmpv4 then
     Result.Error "packet too short for ICMPv4 header"
-  else begin
-    let open Icmpv4_wire in
-    let ty = get_icmpv4_ty buf in
-    let code = get_icmpv4_code buf in
-    let csum = get_icmpv4_csum buf in
-    let subheader = subheader_of_cstruct ty (Cstruct.shift buf 4) in
-    let payload =
-      if Cstruct.len buf > sizeof_icmpv4
-      then Some (Cstruct.shift buf sizeof_icmpv4)
-      else None
-    in
-    Result.Ok { code; ty; csum; subheader; payload }
-  end
+  else Result.Ok () in
+  let check_ty () =
+    match int_to_ty (get_icmpv4_ty buf) with
+    | None -> Result.Error "unrecognized ICMPv4 type"
+    | Some ty -> Result.Ok ty
+  in
+  check_len () >>= check_ty >>= fun ty ->
+  let code = get_icmpv4_code buf in
+  let csum = get_icmpv4_csum buf in
+  let subheader = subheader_of_cstruct ty (Cstruct.shift buf 4) in
+  let payload =
+    if Cstruct.len buf > sizeof_icmpv4
+    then Some (Cstruct.shift buf sizeof_icmpv4)
+    else None
```

Comment:
(not changed in this PR but...) Does this need to be an option type? If we shift in all cases we'll just get an empty cstruct, which is fine right?